python/TestHarness/RaceChecker.py

```diff
@@ -0,0 +1,112 @@
+#* This file is part of the MOOSE framework
+#* https://www.mooseframework.org
+#*
+#* All rights reserved, see COPYRIGHT for full restrictions
+#* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+#*
+#* Licensed under LGPL 2.1, please see LICENSE for details
+#* https://www.gnu.org/licenses/lgpl-2.1.html
+
+import sys
+import itertools
+
+class RaceChecker(object):
+    """docstring for RaceChecker."""
+
+    def __init__(self, all_jobs):
+        """ Create a RaceChecker object that has all the jobs"""
+        self.all_jobs = all_jobs
+
+
+    def findRacePartners(self):
+        """ Collect all the jobs that have race conditions """
+        raceConditionsExist = False
+
+        for job_a, job_b in itertools.combinations(self.all_jobs, 2):
+            for filesInA in job_a.modifiedFiles:
+                for filesInB in job_b.modifiedFiles:
+                    if (filesInA == filesInB):
+                        if not (job_a in job_b.getDownstreamNodes() or job_b in job_a.getDownstreamNodes()):
+                            job_a.racePartners.add(job_b)
+                            job_b.racePartners.add(job_a)
+                            raceConditionsExist = True
+        return raceConditionsExist
```

Comment:
Okay, so I am stuck at this method. There are better ways of going about the nested for loops (use sets instead to perform differences/intersections), but the more important issue is, are we needing job_$.getDownstreamNodes()?

If we do, then I am unclear as to why. From what I understand about this PR, we have made every job have an edge pointing at each other. So theoretically, there should always be downstreams.

However, what is unexpected about it, is that you are using the DAG which gets deconstructed as it passes through the scheduler. So there are never any downstream jobs by the time we get this far. I think We need to use the dag.getOriginalDAG's DAG to perform this operation correctly.

While I have been toying with this method using the original dag, I am still unclear as to when we want to actually detect a race condition.

Some other optimizations this method needs, is whether or not to perform additional work if certain conditions are met (effectively does the same thing)
```python
def findRacePartners(self):
    for job_a, job_b in itertools.combinations(job_group, 2):
        # Job was skipped/never ran
        if job_a.isSkip() or job_b.isSkip():
            continue
        # No matching files
        elif not set(job_a.modifiedFiles).intersection(set(job_b.modifiedFiles)):
            continue
        # job_a and job_b have similar downstream jobs
        elif not set(job_a.getDownstreamNodes()).intersection(set(job_b.getDownstreamNodes())):
            job_a.racePartners.add(job_b)
            job_b.racePartners.add(job_a)
            return True
```
Even though I have optimized this method, I don't think it is correct. But I wanted to give you an idea how it _could_ be optimized.

Reply:
As we discussed, I thought we were operating on the DAG when doing job_a.getDownstreamNodes(). My bad. Rather than edit my comment, I'm replying so as to preserve some history. Some of it is useful :)